source/common/upstream/upstream_impl.cc

```diff
@@ -933,11 +938,21 @@ ClusterImplBase::ClusterImplBase(
           factory_context.singletonManager(), factory_context.dispatcher())) {
   factory_context.setInitManager(init_manager_);
   auto socket_factory = createTransportSocketFactory(cluster, factory_context);
+  auto* raw_factory_pointer = socket_factory.get();
+
   auto socket_matcher = std::make_unique<TransportSocketMatcherImpl>(
       cluster.transport_socket_matches(), factory_context, socket_factory, *stats_scope);
   info_ = std::make_unique<ClusterInfoImpl>(cluster, factory_context.clusterManager().bindConfig(),
                                             runtime, std::move(socket_matcher),
                                             std::move(stats_scope), added_via_api, factory_context);
+
+  if ((info_->features() & ClusterInfoImpl::Features::USE_ALPN) &&
+      !raw_factory_pointer->supportsAlpn()) {
+    throw EnvoyException(
+        fmt::format("ALPN configured for a cluster which has a non-ALPN transport socket: {}",
+                    cluster.DebugString()));
```

Comment:
nit: I would probably print the cluster name here vs. debug string but up to you.